subsys/usb/device_next/usbd_msg.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <zephyr/usb/usbd.h>
+
+#include "usbd_device.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbd_msg, CONFIG_USBD_LOG_LEVEL);
+
+static void msg_work_handler(struct k_work *work);
+static K_WORK_DELAYABLE_DEFINE(msg_work, msg_work_handler);
+K_FIFO_DEFINE(msg_queue);
+
+struct usbd_msg_pkt {
+	sys_snode_t node;
+	struct usbd_contex *ctx;
+	struct usbd_msg msg;
+};
+
+K_MEM_SLAB_DEFINE_STATIC(usbd_msg_slab, sizeof(struct usbd_msg_pkt),
+			 CONFIG_USBD_MAX_MESSAGES, sizeof(void *));
+
+static ALWAYS_INLINE int submit_new_work(struct usbd_msg_pkt *const m_pkt)
+{
+	k_fifo_put(&msg_queue, m_pkt);
+	return k_work_schedule(&msg_work, K_NO_WAIT);
+}
+
+static inline struct usbd_msg_pkt *usbd_alloc_msg(struct usbd_contex *const ctx)
+{
+	struct usbd_msg_pkt *m_pkt;
+
+	if (k_mem_slab_alloc(&usbd_msg_slab, (void **)&m_pkt, K_NO_WAIT)) {
```

Comment:
not blocking here?

Reply:
We cannot block, and should not, described in the issue.